datacube.py

```diff
@@ -0,0 +1,382 @@
+
+    Returns:
+    - tuple: A tuple containing the STAC catalog, Sentinel-2 items, Sentinel-2 GeoDataFrame, and the bounding box (BBOX).
+    """
+
+    CENTROID = aoi.centroid
+    BBOX = aoi.bounds
+
+    geom_CENTROID = Point(CENTROID.x, CENTROID.y)  # Create point geom object from centroid
+
+    catalog = pystac_client.Client.open(STAC_API, modifier=pc.sign_inplace)
+
+    search: pystac_client.item_search.ItemSearch = catalog.search(
+        filter_lang="cql2-json",
+        filter={
+            "op": "and",
+            "args": [
+                {
+                    "op": "s_intersects",
+                    "args": [{"property": "geometry"}, geom_CENTROID.__geo_interface__],
+                },
+                {"op": "anyinteracts", "args": [{"property": "datetime"}, week]},
+                {"op": "=", "args": [{"property": "collection"}, "sentinel-2-l2a"]},
+                {"op": "<=", "args": [{"property": "eo:cloud_cover"}, 50]},
+                {"op": "<=", "args": [{"property": "s2:nodata_pixel_percentage"}, 20]},
+            ],
+        },
+    )
+
+    s2_items = search.get_all_items()
+    print(f"Found {len(s2_items)} Sentinel-2 items")
+
+    s2_items_gdf = gpd.GeoDataFrame.from_features(s2_items.to_dict())
+
+    BBOX = s2_items_gdf.iloc[0].geometry.bounds
+
+    return catalog, s2_items, s2_items_gdf, BBOX
+
+
+def search_sentinel1(BBOX, catalog, week):
+    """
+    Search for Sentinel-1 items within a given bounding box (BBOX), STAC catalog, and week.
+
+    Parameters:
+    - BBOX (tuple): Bounding box coordinates in the format (minx, miny, maxx, maxy).
+    - catalog (pystac.Catalog): STAC catalog containing Sentinel-1 items.
+    - week (str): The week in the format 'start_date/end_date'.
+
+    Returns:
+    - tuple: A tuple containing Sentinel-1 items and Sentinel-1 GeoDataFrame.
+    """
+
+    geom_BBOX = box(*BBOX)  # Create poly geom object from the bbox
+
+    search: pystac_client.item_search.ItemSearch = catalog.search(
+        filter_lang="cql2-json",
+        filter={
+            "op": "and",
+            "args": [
+                {
+                    "op": "s_intersects",
+                    "args": [{"property": "geometry"}, geom_BBOX.__geo_interface__],
+                },
+                {"op": "anyinteracts", "args": [{"property": "datetime"}, week]},
+                {"op": "<=", "args": [{"property": "sat:orbit_state"}, "descending"]},
+                {"op": "=", "args": [{"property": "collection"}, "sentinel-1-rtc"]},
+            ],
+        },
+    )
+    s1_items = search.item_collection()
+    print(f"Found {len(s1_items)} Sentinel-1 items")
+
+    s1_gdf = gpd.GeoDataFrame.from_features(s1_items.to_dict())
+    # TODO: Find a way to get minimal number of S1 tiles to cover the entire S2 bbox
+    s1_gdf["overlap"] = s1_gdf.intersection(box(*BBOX)).area
+    s1_gdf.sort_values(by="overlap", inplace=True)
```

Comment:
Pick tiles from ascending or descending capture based on their overlap with the S2 bbox.
```
state = (s1_gdf[["sat:orbit_state", "overlap"]]
             .groupby(["sat:orbit_state"])
             .sum()
             .sort_values(by="overlap", ascending=False)
             .index[0])
```
`state = ascending OR descending`. 

Filter the s1_gdf by `state`: `s1_gdf = s1_gdf[s1_gdf["sat:orbit_state"] == state]`

Reply:
Implemented